_posts/2019-01-03-elixir-kaffe-codealong.md

```diff
@@ -0,0 +1,204 @@
+# Connecting Elixir to Kafka with Kaffe: A codealong
+_The most basic example that exists_
+
+If we want to use the popular messaging system Kafka with our Elixir projects, we have a few wrappers we can choose from. One of them, [Kaffe](https://github.com/spreedly/kaffe), proved difficult to troubleshoot. Just try googling "Elixir Kaffe" yourself.
+
+In this codealong we'll build the barest of simple Elixir applications and use Kaffe to connect it to a locally running Kafka server. Later we'll cover a couple of variations to connect a dockerized Kafka server or an umbrella elixir app.
+
+This post assumes basic knowledge of Elixir and no knowledge of Kafka or Kaffe. Here is the repo with the full project: [Elixir Kaffe Codealong](https://github.com/meryldakin/elixir_kaffe_codealong).
+
+## What is Kafka, briefly?
+Kafka is a messaging system. It does essentially three things:
+1. Receives messages from applications
+2. Keeps those messages in the order they were receieved in
+3. Allows other applications to read those messages in order
+
+Learn a lot more here: [Kafka in a Nutshell](https://sookocheff.com/post/kafka/kafka-in-a-nutshell/)
+
+## Basic Kafka terminology in simple terms
+- **consumer:** what is receiving messages from Kafka
+- **producer:** what is sending messages to Kafka
+- **topic:** a way to organize messages and allow consumers to only subscribe to the ones they want to receive
+- **partition:** allows a topic to be split among multiple machines and retain the same data so that more than one consumer can read from a single topic at a time
+- **leader/replica:** these are types of partitions. There is one leader and multiple replicas. The leader makes sure the replicas have the same and newest data. If the leader fails, a replica will take over as leader.
+- **offset:** the unique identifier of a message that keeps its order within Kafka
+
+## Codealong: basic Elixir app & Kafka running locally
+
+### Set up Kafka Server
+Follow the first two steps of the [quickstart instructions](http://kafka.apache.org/documentation/#quickstart) from Apache Kafka:
+1. [Download the code](https://www.apache.org/dyn/closer.cgi?path=/kafka/2.1.0/kafka_2.11-2.1.0.tgz)
+2. Start the servers
+Zookeeper
+`bin/zookeeper-server-start.sh config/zookeeper.properties`
+Kafka
+`bin/kafka-server-start.sh config/server.properties`
+
+### Set up Elixir App
+
+**1. Start new project**
+`mix new elixir_kaffe_codealong`
+
+**2. Configure kaffe**
+**2.a:** In `mix.exs` add `:kaffe` to the list of extra applications:
```

Comment:
I don't _think_ we need to manage these bullets ourselves, this is built into Markdown: https://gist.github.com/roachhd/779fa77e9b90fe945b0c

It might be better to rely on Markdown for the lists and sublists so they're consist with lists elsewhere in blogs and lessons.